File: src/graphql/mutations/billing/rateCards.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// RATE CARD MUTATIONS
// ============================================================================

/**
 * Creates a new rate card.
 * @param payload The input data for creating the rate card.
 */
export const createRateCard = graphql(`
  mutation CreateRateCard($payload: CreateRateCardInput!) {
    billing {
      createRateCard(payload: $payload) {
        id
        name
        serviceType
        isActive
        validFrom
        validTo
        description
        createdAt
        updatedAt
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the name of a rate card.
 * @param id The UUID of the rate card to update.
 * @param name The new name.
 */
export const updateRateCardName = graphql(`
  mutation UpdateRateCardName($id: UUID!, $name: String!) {
    billing {
      updateRateCardName(id: $id, name: $name) {
        id
        name
        serviceType
        isActive
        validFrom
        validTo
        description
        createdAt
        updatedAt
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the service type of a rate card.
 * @param id The UUID of the rate card to update.
 * @param serviceType The new service type.
 */
export const updateRateCardServiceType = graphql(`
  mutation UpdateRateCardServiceType($id: UUID!, $serviceType: ServiceTypeEnum!) {
    billing {
      updateRateCardServiceType(id: $id, serviceType: $serviceType) {
        id
        name
        serviceType
        isActive
        validFrom
        validTo
        description
        createdAt
        updatedAt
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the active status of a rate card.
 * @param id The UUID of the rate card to update.
 * @param isActive The new active status.
 */
export const updateRateCardIsActive = graphql(`
  mutation UpdateRateCardIsActive($id: UUID!, $isActive: Boolean) {
    billing {
      updateRateCardIsActive(id: $id, isActive: $isActive) {
        id
        name
        serviceType
        isActive
        validFrom
        validTo
        description
        createdAt
        updatedAt
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the valid from date of a rate card.
 * @param id The UUID of the rate card to update.
 * @param validFrom The new valid from date.
 */
export const updateRateCardValidFrom = graphql(`
  mutation UpdateRateCardValidFrom($id: UUID!, $validFrom: NaiveDate!) {
    billing {
      updateRateCardValidFrom(id: $id, validFrom: $validFrom) {
        id
        name
        serviceType
        isActive
        validFrom
        validTo
        description
        createdAt
        updatedAt
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the valid to date of a rate card.
 * @param id The UUID of the rate card to update.
 * @param validTo The new valid to date.
 */
export const updateRateCardValidTo = graphql(`
  mutation UpdateRateCardValidTo($id: UUID!, $validTo: NaiveDate) {
    billing {
      updateRateCardValidTo(id: $id, validTo: $validTo) {
        id
        name
        serviceType
        isActive
        validFrom
        validTo
        description
        createdAt
        updatedAt
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the description of a rate card.
 * @param id The UUID of the rate card to update.
 * @param description The new description.
 */
export const updateRateCardDescription = graphql(`
  mutation UpdateRateCardDescription($id: UUID!, $description: String) {
    billing {
      updateRateCardDescription(id: $id, description: $description) {
        id
        name
        serviceType
        isActive
        validFrom
        validTo
        description
        createdAt
        updatedAt
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the created by user ID of a rate card.
 * @param id The UUID of the rate card to update.
 * @param createdByUserId The new created by user ID.
 */
export const updateRateCardCreatedByUserId = graphql(`
  mutation UpdateRateCardCreatedByUserId($id: UUID!, $createdByUserId: UUID) {
    billing {
      updateRateCardCreatedByUserId(id: $id, createdByUserId: $createdByUserId) {
        id
        name
        serviceType
        isActive
        validFrom
        validTo
        description
        createdAt
        updatedAt
        createdByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Removes a rate card by its ID.
 * @param id The UUID of the rate card to remove.
 */
export const removeRateCard = graphql(`
  mutation RemoveRateCard($id: UUID!) {
    billing {
      removeRateCard(id: $id)
    }
  }
`);

File: services/graphql-billing/src/mutation/rate_cards.rs
use async_graphql::{Context, InputObject, Object};
use chrono::NaiveDate;
use uuid::Uuid;

use crate::models::{enums::ServiceTypeEnum, rate_cards};

#[derive(Debug, Clone, InputObject)]
pub struct CreateRateCardInput {
    pub name: String,
    pub service_type: ServiceTypeEnum,
    pub is_active: Option<bool>,
    pub valid_from: NaiveDate,
    pub valid_to: Option<NaiveDate>,
    pub description: Option<String>,
    pub created_by_user_id: Option<Uuid>,
}

#[derive(Debug, Clone, Default)]
pub struct RateCardsMutation;

#[Object(name = "BillingRateCardsMutation")]
impl RateCardsMutation {
    async fn create_rate_card(
        &self,
        ctx: &Context<'_>,
        payload: CreateRateCardInput,
    ) -> async_graphql::Result<rate_cards::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, rate_cards::Model>(
            "insert into billing.rate_cards (name, service_type, is_active, valid_from, valid_to, description, created_by_user_id) values ($1, $2, $3, $4, $5, $6, $7) returning *",
        )
        .bind(payload.name)
        .bind(payload.service_type)
        .bind(payload.is_active)
        .bind(payload.valid_from)
        .bind(payload.valid_to)
        .bind(payload.description)
        .bind(payload.created_by_user_id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_rate_card_name(
        &self,
        ctx: &Context<'_>,
        name: String,
        id: Uuid,
    ) -> async_graphql::Result<rate_cards::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, rate_cards::Model>(
            "update billing.rate_cards set name = $1 where id = $2 returning *",
        )
        .bind(name)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_rate_card_service_type(
        &self,
        ctx: &Context<'_>,
        service_type: ServiceTypeEnum,
        id: Uuid,
    ) -> async_graphql::Result<rate_cards::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, rate_cards::Model>(
            "update billing.rate_cards set service_type = $1 where id = $2 returning *",
        )
        .bind(service_type)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_rate_card_is_active(
        &self,
        ctx: &Context<'_>,
        is_active: Option<bool>,
        id: Uuid,
    ) -> async_graphql::Result<rate_cards::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, rate_cards::Model>(
            "update billing.rate_cards set is_active = $1 where id = $2 returning *",
        )
        .bind(is_active)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_rate_card_valid_from(
        &self,
        ctx: &Context<'_>,
        valid_from: NaiveDate,
        id: Uuid,
    ) -> async_graphql::Result<rate_cards::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, rate_cards::Model>(
            "update billing.rate_cards set valid_from = $1 where id = $2 returning *",
        )
        .bind(valid_from)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_rate_card_valid_to(
        &self,
        ctx: &Context<'_>,
        valid_to: Option<NaiveDate>,
        id: Uuid,
    ) -> async_graphql::Result<rate_cards::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, rate_cards::Model>(
            "update billing.rate_cards set valid_to = $1 where id = $2 returning *",
        )
        .bind(valid_to)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_rate_card_description(
        &self,
        ctx: &Context<'_>,
        description: Option<String>,
        id: Uuid,
    ) -> async_graphql::Result<rate_cards::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, rate_cards::Model>(
            "update billing.rate_cards set description = $1 where id = $2 returning *",
        )
        .bind(description)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_rate_card_created_by_user_id(
        &self,
        ctx: &Context<'_>,
        created_by_user_id: Option<Uuid>,
        id: Uuid,
    ) -> async_graphql::Result<rate_cards::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, rate_cards::Model>(
            "update billing.rate_cards set created_by_user_id = $1 where id = $2 returning *",
        )
        .bind(created_by_user_id)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn remove_rate_card(&self, ctx: &Context<'_>, id: Uuid) -> async_graphql::Result<String> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query("delete from billing.rate_cards where id = $1")
            .bind(id)
            .execute(db)
            .await?;

        if result.rows_affected() != 1 {
            return Err(async_graphql::Error::new("Unable to remove rate card"));
        }

        Ok("Rate card removed successfully".into())
    }
}
